docs/development/ceps/proposed/cep-002.rst

```diff
@@ -0,0 +1,204 @@
+where each ``ArrayEventContainer`` is composed of one to many ``TelescopeEvents`` containing
+all telescope-wise information for all data levels.
+
+The ``ArrayEventContainer`` should also be renamed to ``SubarrayEventContainer``, to match with other naming
+patterns in ctapipe, such as the ``SubarrayDescription``, making it clear that the array is split
+into multiple subarrays, each observing their own observation block.
+
+The main structure after the change will look like this (Container-suffix left out for readability):
+
+.. code-block::
+
+   SubarrayEvent
+   - index: SubarrayEventIndex
+   - trigger: SubarrayTrigger
+   - simulation: SimulatedShower
+   - dl0: DL0Subarray
+   - dl1: DL1Subarray
+   - dl2: DL2Subarray
+   - dl3: DL3Subarray
+   - tel: Map[tel_id -> TelescopeEvent]
+
+   TelescopeEvent
+   - index: TelescopeEventIndex
+   - trigger: TelescopeTrigger
+   - simulation: TelescopeSimulation
+   - r0: R0Telescope
+   - r1: R1Telescope
+   - dl0: DL0Telescope
+   - dl1: DL1Telescope
+   - dl2: DL2Telescope
+   - dl3: DL3Telescope
+
+Which each data level container having specific fields and/or subcontainers including monitoring
+information (interpolated / chosen for that specific event).
+
+
+Advantages of the new structure
+===============================
+
+The new proposed scheme makes it easier to parallelize over array events and move loops
+over telescopes out of code paths that only deal with a single telescope.
+E.g. in the ``CameraCalibrator``, ``ImageProcessor`` and more classes,
+we currently have to provide the ``ArrayEventContainer``,
+so that higher data levels can be filled from lower data levels, although only one telescope
+is processed at a time.
+
+In the current scheme (simplified from ctapipe-process) it looks like this:
+
+.. code-block:: python
+
+   for array_event in source:
+       calibrator(array_event)
+       # calibrator internally has two hidden loops like this:
+       # for tel_id, r1 in array_event.r1.tel.items():
+       #     calibrate r1 to dl0
+       # for tel_id, dl0 in array_event.dl0.tel.items():
+       #     calibrate dl0 to dl1
+       
+       image_processor(array_event)
+       # image processor also has an internal loop over the telescope events
+       # for tel_id, dl1 in array_event.dl1.tel.items():
+       #     image cleaning and parametrization
+       
+       shower_processor(array_event)
+
+This looks simple, but there are hidden loops over the telescopes in both the ``CameraCalibrator``
+and the ``ImageProcessor`` here, although both of these do not access any subarray-wide data at all.
+
+Using the new structure, these classes will get a single ``TelescopeEventContainer`` and the loop
+can to be moved outside those classes to a single place:
+
+.. code-block:: python
+
+   for array_event in source:
+       for telescope_event in array_event.tel.values():
+           calibrator(telescope_event)
```

Comment:
Do I understand it correctly if I interpret the example code as saying the calibrator in the new structure does not need either of the two loops present in the old example? 

Eliminating _two_ loops over all events seems like a signification savings, and it is worth spelling it out properly in the text.

Reply:
Nothing here changes the amount of work done. Since the calibration is quite work-intensive, I doubt it matters for performance.

Reply:
Nothing here changes the amount of work done, the difference will only be the python overhead of the loops.

Since the calibration is quite work-intensive, I doubt it matters for performance.

Reply:
Essentially we change from:


```
for array_event in source:
    for tel_event in array_event:
        step1(tel_event)
    for tel_event in array_event:
        step2(tel_event)
```


to


```
for array_event in source:
    for tel_event in array_event:
        step1(tel_event)
        step2(tel_event)
```


Ignoring where the loops are (inside or outside of `CameraCalibrator`.